server/src/main/java/org/opensearch/index/shard/IndexShard.java

```diff
@@ -1697,7 +1697,8 @@ ReplicationCheckpoint computeReplicationCheckpoint(SegmentInfos segmentInfos) th
         }
         final ReplicationCheckpoint latestReplicationCheckpoint = getLatestReplicationCheckpoint();
         if (latestReplicationCheckpoint.getSegmentInfosVersion() == segmentInfos.getVersion()
-            && latestReplicationCheckpoint.getSegmentsGen() == segmentInfos.getGeneration()) {
+            && latestReplicationCheckpoint.getSegmentsGen() == segmentInfos.getGeneration()
+            && latestReplicationCheckpoint.getPrimaryTerm() == getOperationPrimaryTerm()) {
```

Comment:
This would work if the failover happens with following order -
1. operation primary term is updated
2. activate primary mode

Is it the same order during failover?

Reply:
yes 

https://github.com/opensearch-project/OpenSearch/blob/30aa8bec12220e528d7fd635c78ea4d76c686917/server/src/main/java/org/opensearch/index/shard/IndexShard.java#L4062-L4067

the onBlocked.run() will activate primarymode